core/services/ocr2/plugins/threshold/decryption_queue.go

```diff
@@ -0,0 +1,226 @@
+package decryption_queue
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type CiphertextId = []byte
+
+type ThresholdDecryptor interface {
+	Decrypt(ctx context.Context, ciphertextId CiphertextId, ciphertext []byte) ([]byte, error)
+}
+
+// This interface will be replaced by the Threshold Decryption Plugin when it is ready
+type DecryptionQueuingService interface {
+	GetRequests(requestCountLimit uint32, totalBytesLimit uint32) []DecryptionRequest
+	GetCiphertext(ciphertextId CiphertextId) ([]byte, error)
+	ResultReady(ciphertextId CiphertextId, plaintext []byte)
+}
+
+type DecryptionRequest struct {
+	ciphertextId CiphertextId
+	ciphertext   []byte
+}
+
+type pendingRequest struct {
+	chPlaintext chan []byte
+	ciphertext  []byte
+}
+
+type completedRequest struct {
+	plaintext []byte
+	timer     *time.Timer
+}
+
+type decryptionQueue struct {
+	maxQueueLength                uint32
+	maxCiphertextBytes            uint32
+	completedRequestsCacheTimeout time.Duration
+	pendingRequestQueue           []CiphertextId
+	pendingRequests               map[string]pendingRequest
+	completedRequests             map[string]completedRequest
+	mu                            sync.RWMutex
+	lggr                          logger.Logger
+}
+
+var (
+	_ ThresholdDecryptor       = &decryptionQueue{}
+	_ DecryptionQueuingService = &decryptionQueue{}
+	_ job.ServiceCtx           = &decryptionQueue{}
+)
+
+func NewThresholdDecryptor(maxQueueLength uint32, maxCiphertextBytes uint32, completedRequestsCacheTimeoutMs uint64, lggr logger.Logger) ThresholdDecryptor {
```

Comment:
A common pattern/proverb in Go is to "accept interfaces, return structs". In this case you could consider exporting `decryptionQueue` as `DecryptionQueue` and returning a pointer to the concrete type:
```suggestion
func NewDecryptionQueue(maxQueueLength uint32, maxCiphertextBytes uint32, completedRequestsCacheTimeoutMs uint64, lggr logger.Logger) *DecryptionQueue {
```
It keeps the API more flexible (namely you can add more methods later in a compatible way), but the immediate benefit here would be eliminating all this test boilerplate:
```diff
-dq, ok := td.(*decryptionQueue)
-require.True(t, ok)
```

Reply:
nit/ IMHO it is more clear to accept the pure `time.Duration` that is used internally than a unit-suffixed type to convert:
```suggestion
func NewThresholdDecryptor(maxQueueLength uint32, maxCiphertextBytes uint32, completedRequestsCacheTimeout time.Duration, lggr logger.Logger) ThresholdDecryptor {
```

Reply:
> nit/ IMHO it is more clear to accept the pure `time.Duration` that is used internally than a unit-suffixed type to convert:

Thanks! I will leave this as feedback on threshold decryption plugin design that was created in collaboration with Research to ensure we are in agreement on this interface change.

Reply:
> A common pattern/proverb in Go is to "accept interfaces, return structs". In this case you could consider exporting `decryptionQueue` as `DecryptionQueue` and returning a pointer to the concrete type:
> 
> It keeps the API more flexible (namely you can add more methods later in a compatible way), but the immediate benefit here would be eliminating all this test boilerplate:
> 
> ```diff
> -dq, ok := td.(*decryptionQueue)
> -require.True(t, ok)
> ```

Thank you! This slightly conflicts with @pinebit 's recommendation.  I will discuss with him to resolve this.

Reply:
Oh is this constructor's method signature used somewhere else? This sort of thing is normally just an implementation detail.